### PR TITLE
(fix)chaos_executor: ensure executor waits for the experiment chaos duration before completion

### LINKED ISSUE #953: Executor terminates before experiment job completion
**What happened**:

- Currently, the chaos executor monitors experiment job for a fixed period of 300s, post which it terminates ungracefully if the experiment is still in progress

- The experiments have a tunable called "TOTAL_CHAOS_DURATION" which can exceed the above timeout.

**What you expected to happen**:

- The executor is expected to wait/monitor the experiment until the chaos duration for completion. 

**How to reproduce it (as minimally and precisely as possible)**:

- Run an experiment with TOTAL_CHAOS_DURATION > 300s

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by amitbhatt818**
Signed-off-by: Amit Bhatt <amitbhatt818@gmail.com>

**What this PR does / why we need it**:

- This PR ensures that the chaos executor monitors the experiment for the period of  ```TOTAL_CHAOS_DURATION``` (adds some default period of 500s to it) 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #953 

**Checklist**

* [x] Does this PR have a corresponding GitHub issue?
* [ ] Have you included relevant README for the chaoslib/experiment with details?
* [ ] Have you added debug messages where necessary? 
* [ ] Have you added task comments where necessary? 
* [ ] Have you tested the changes for possible failure conditions?
* [ ] Have you provided the positive & negative test logs for the litmusbook execution?
* [ ] Does the litmusbook ensure idempotency of cluster state?, i.e., is cluster restored to original state?
* [ ] Have you used non-shell/command modules for Kubernetes tasks?
* [ ] Have you (jinja) templatized custom scripts that is run by the litmusbook, if any? 
* [ ] Have you (jinja) templatized Kubernetes deployment manifests used by the litmusbook, if any?
* [ ] Have you reused/created util functions instead of repeating tasks in the litmusbook?
* [ ] Do the artifacts follow the appropriate directory structure? 
* [ ] Have you isolated storage (eg: OpenEBS) specific implementations, checks? 
* [ ] Have you isolated platform (eg: baremetal kubeadm/openshift/aws/gcloud) specific implementations, checks?  
* [ ] Are the ansible facts well defined? Is the scope explicitly set for playbook & included utils?
* [ ] Have you ensured minimum/careful usage of shell utilities (awk, grep, sed, cut, xargs etc.,)?
* [ ] Can the limtusbook be executed both from within & outside a container (configurable paths, no hardcode)?
* [ ] Can you suggest the minimal resource requirements for the litmusbook execution?
* [ ] Does the litmusbook job artifact carry comments/default options/range for the ENV tunables?
* [ ] Has the litmusbooks been linted? 

**Special notes for your reviewer**:

### CODE REVIEW

File: executor/executor.yml [hunk 1]
```diff
@@ -154,7 +154,14 @@
 
 - include: experiment_configmap.yml
   when:  configMap_available != '' or  secret_available != ''
-  
+
+- name: Fetching monitoring time from chaos experiment job
+  shell: >
+    kubectl get job {{ job_name }} -n {{c_app_ns}} -ojsonpath="{range .spec.template.spec.containers[*]}{.env[?(@.name=='TOTAL_CHAOS_DURATION')].value}"
```
Comment (ksatchit) [inline executor/executor.yml]: :+1:

Comment (amitbhatt818) [reply executor/executor.yml]: Thank-you @ksatchit

**Comment by ksatchit** [comment]
**Logs-**

```
PLAY [localhost] ***************************************************************
2019-11-22T11:24:54.472558 (delta: 0.038055)         elapsed: 0.038055 ******** 
=============================================================================== 

TASK [Gathering Facts] *********************************************************
2019-11-22T11:24:54.482429 (delta: 0.00985)         elapsed: 0.047926 ********* 
ok: [127.0.0.1]

TASK [debug] *******************************************************************
2019-11-22T11:24:57.959629 (delta: 3.477182)         elapsed: 3.525126 ******** 
ok: [127.0.0.1] => {
    "msg": [
        "List of Chaos Experiments: [u'pod-delete']", 
        "Each experiment will be launched & monitored by the executor", 
        "Beginning Execution, view experiment logs on respective job"
    ]
}

TASK [debug] *******************************************************************
2019-11-22T11:24:58.023008 (delta: 0.063357)         elapsed: 3.588505 ******** 
ok: [127.0.0.1] => {
    "msg": [
        "c_rand value is: ", 
        "c_max value is: "
    ]
}

TASK [include] *****************************************************************
2019-11-22T11:24:58.085201 (delta: 0.062161)         elapsed: 3.650698 ******** 
included: /executor/executor.yml for 127.0.0.1

TASK [Verify if ChaosExperiment CR is available] *******************************
2019-11-22T11:24:58.204432 (delta: 0.119185)         elapsed: 3.769929 ******** 
changed: [127.0.0.1]

TASK [check availabilty of experimentStatuses] *********************************
2019-11-22T11:24:59.013600 (delta: 0.809126)         elapsed: 4.579097 ******** 
changed: [127.0.0.1]

TASK [Obtain the env param list length from spec of experiment] ****************
2019-11-22T11:24:59.554255 (delta: 0.540621)         elapsed: 5.119752 ******** 
changed: [127.0.0.1]

TASK [set_fact] ****************************************************************
2019-11-22T11:25:00.100090 (delta: 0.545779)         elapsed: 5.665587 ******** 
ok: [127.0.0.1]

TASK [Obtain the labels from the experiment] ***********************************
2019-11-22T11:25:00.165135 (delta: 0.065011)         elapsed: 5.730632 ******** 
changed: [127.0.0.1]

TASK [Fetching image from the chaosexperiment] *********************************
2019-11-22T11:25:00.909561 (delta: 0.744382)         elapsed: 6.475058 ******** 
changed: [127.0.0.1]

TASK [Obtain the args from the experiment] *************************************
2019-11-22T11:25:01.450088 (delta: 0.5405)         elapsed: 7.015585 ********** 
changed: [127.0.0.1]

TASK [Check Availability of configmaps] ****************************************
2019-11-22T11:25:02.065778 (delta: 0.615616)         elapsed: 7.631275 ******** 
changed: [127.0.0.1]

TASK [Check Availability of secrets] *******************************************
2019-11-22T11:25:02.603077 (delta: 0.537269)         elapsed: 8.168574 ******** 
changed: [127.0.0.1]

TASK [Record availability of configmaps and secrets] ***************************
2019-11-22T11:25:03.166297 (delta: 0.563179)         elapsed: 8.731794 ******** 
ok: [127.0.0.1]

TASK [include] *****************************************************************
2019-11-22T11:25:03.221138 (delta: 0.054816)         elapsed: 8.786635 ******** 
included: /executor/experiment_env_getter.yml for 127.0.0.1 => (item=0)
included: /executor/experiment_env_getter.yml for 127.0.0.1 => (item=1)
included: /executor/experiment_env_getter.yml for 127.0.0.1 => (item=2)
included: /executor/experiment_env_getter.yml for 127.0.0.1 => (item=3)

TASK [Get ENV names for Chaos Experiment] **************************************
2019-11-22T11:25:03.354598 (delta: 0.132334)         elapsed: 8.920095 ******** 
changed: [127.0.0.1]

TASK [Get ENV values for Chaos Experiment] *************************************
2019-11-22T11:25:03.908476 (delta: 0.553803)         elapsed: 9.473973 ******** 
changed: [127.0.0.1]

TASK [append envNameList] ******************************************************
2019-11-22T11:25:04.498802 (delta: 0.590269)         elapsed: 10.064299 ******* 
ok: [127.0.0.1]

TASK [append envValueList] *****************************************************
2019-11-22T11:25:04.567252 (delta: 0.068421)         elapsed: 10.132749 ******* 
ok: [127.0.0.1]

TASK [Get ENV names for Chaos Experiment] **************************************
2019-11-22T11:25:04.620647 (delta: 0.053368)         elapsed: 10.186144 ******* 
changed: [127.0.0.1]

TASK [Get ENV values for Chaos Experiment] *************************************
2019-11-22T11:25:05.168113 (delta: 0.54744)         elapsed: 10.73361 ********* 
changed: [127.0.0.1]

TASK [append envNameList] ******************************************************
2019-11-22T11:25:05.723837 (delta: 0.555667)         elapsed: 11.289334 ******* 
ok: [127.0.0.1]

TASK [append envValueList] *****************************************************
2019-11-22T11:25:05.812553 (delta: 0.088674)         elapsed: 11.37805 ******** 
ok: [127.0.0.1]

TASK [Get ENV names for Chaos Experiment] **************************************
2019-11-22T11:25:05.873339 (delta: 0.06073)         elapsed: 11.438836 ******** 
changed: [127.0.0.1]

TASK [Get ENV values for Chaos Experiment] *************************************
2019-11-22T11:25:06.404822 (delta: 0.531455)         elapsed: 11.970319 ******* 
changed: [127.0.0.1]

TASK [append envNameList] ******************************************************
2019-11-22T11:25:06.987887 (delta: 0.58304)         elapsed: 12.553384 ******** 
ok: [127.0.0.1]

TASK [append envValueList] *****************************************************
2019-11-22T11:25:07.038480 (delta: 0.050565)         elapsed: 12.603977 ******* 
ok: [127.0.0.1]

TASK [Get ENV names for Chaos Experiment] **************************************
2019-11-22T11:25:07.091758 (delta: 0.053253)         elapsed: 12.657255 ******* 
changed: [127.0.0.1]

TASK [Get ENV values for Chaos Experiment] *************************************
2019-11-22T11:25:07.669466 (delta: 0.577636)         elapsed: 13.234963 ******* 
changed: [127.0.0.1]

TASK [append envNameList] ******************************************************
2019-11-22T11:25:08.205986 (delta: 0.536492)         elapsed: 13.771483 ******* 
ok: [127.0.0.1]

TASK [append envValueList] *****************************************************
2019-11-22T11:25:08.263186 (delta: 0.056298)         elapsed: 13.828683 ******* 
ok: [127.0.0.1]

TASK [Obtain the env param list length from spec of Engine] ********************
2019-11-22T11:25:08.322982 (delta: 0.059767)         elapsed: 13.888479 ******* 
changed: [127.0.0.1]

TASK [include] *****************************************************************
2019-11-22T11:25:08.890750 (delta: 0.56771)         elapsed: 14.456247 ******** 

TASK [merging appinfo from chaosengine] ****************************************
2019-11-22T11:25:08.934950 (delta: 0.044163)         elapsed: 14.500447 ******* 
ok: [127.0.0.1]

TASK [Fetching env from dictionary] ********************************************
2019-11-22T11:25:09.000313 (delta: 0.065336)         elapsed: 14.56581 ******** 
ok: [127.0.0.1] => (item={'value': u'', 'key': u'APP_KIND'})
ok: [127.0.0.1] => (item={'value': u'', 'key': u'LIB'})
ok: [127.0.0.1] => (item={'value': u'5', 'key': u'CHAOS_INTERVAL'})
ok: [127.0.0.1] => (item={'value': u'default', 'key': u'ANSIBLE_STDOUT_CALLBACK'})
ok: [127.0.0.1] => (item={'value': u'app=nginx', 'key': u'APP_LABEL'})
ok: [127.0.0.1] => (item={'value': u'default', 'key': u'APP_NAMESPACE'})
ok: [127.0.0.1] => (item={'value': u'15', 'key': u'TOTAL_CHAOS_DURATION'})
ok: [127.0.0.1] => (item={'value': u'engine-nginx', 'key': u'CHAOSENGINE'})

TASK [display env lists] *******************************************************
2019-11-22T11:25:09.130463 (delta: 0.130117)         elapsed: 14.69596 ******** 
ok: [127.0.0.1] => {
    "msg": "namelist: [u'APP_NAMESPACE', u'APP_LABEL', u'APP_KIND', u'CHAOSENGINE', u'ANSIBLE_STDOUT_CALLBACK', u'TOTAL_CHAOS_DURATION', u'CHAOS_INTERVAL', u'LIB'], valuelist: [u'default', u'app=nginx', u'', u'engine-nginx', u'default', u'15', u'5', u''], maplist:  --env=APP_KIND= --env=LIB= --env=CHAOS_INTERVAL=5 --env=ANSIBLE_STDOUT_CALLBACK=default --env=APP_LABEL=app=nginx --env=APP_NAMESPACE=default --env=TOTAL_CHAOS_DURATION=15 --env=CHAOSENGINE=engine-nginx"
}

TASK [Generate a simple random sleep duration] *********************************
2019-11-22T11:25:09.187455 (delta: 0.056953)         elapsed: 14.752952 ******* 
skipping: [127.0.0.1]

TASK [Generate a random sleep duration based on total chaos duration] **********
2019-11-22T11:25:09.226222 (delta: 0.038742)         elapsed: 14.791719 ******* 
skipping: [127.0.0.1]

TASK [Sleep for random duration before triggering experiments] *****************
2019-11-22T11:25:09.265577 (delta: 0.039287)         elapsed: 14.831074 ******* 
skipping: [127.0.0.1]

TASK [debug] *******************************************************************
2019-11-22T11:25:09.326075 (delta: 0.060464)         elapsed: 14.891572 ******* 
ok: [127.0.0.1] => {
    "msg": "Triggering chaos experiment: pod-delete "
}

TASK [genarate random string] **************************************************
2019-11-22T11:25:09.379356 (delta: 0.053252)         elapsed: 14.944853 ******* 
changed: [127.0.0.1]

TASK [Register the Job Name for further use] ***********************************
2019-11-22T11:25:09.874265 (delta: 0.494842)         elapsed: 15.439762 ******* 
ok: [127.0.0.1]

TASK [Run the chaos experiment job] ********************************************
2019-11-22T11:25:09.926684 (delta: 0.052371)         elapsed: 15.492181 ******* 
changed: [127.0.0.1]

TASK [include] *****************************************************************
2019-11-22T11:25:10.530522 (delta: 0.603786)         elapsed: 16.096019 ******* 
skipping: [127.0.0.1]

TASK [Fetching monitoring time from chaos experiment job] **********************
2019-11-22T11:25:10.581211 (delta: 0.050569)         elapsed: 16.146708 ******* 
changed: [127.0.0.1]

TASK [Monitoring the litmus chaos job for completion] **************************
2019-11-22T11:25:11.238129 (delta: 0.656719)         elapsed: 16.803626 ******* 
FAILED - RETRYING: Monitoring the litmus chaos job for completion (515 retries left).
FAILED - RETRYING: Monitoring the litmus chaos job for completion (514 retries left).
FAILED - RETRYING: Monitoring the litmus chaos job for completion (513 retries left).
FAILED - RETRYING: Monitoring the litmus chaos job for completion (512 retries left).
changed: [127.0.0.1]

TASK [Fetching verdict from the chaos result CR] *******************************
2019-11-22T11:25:54.067225 (delta: 42.828843)         elapsed: 59.632722 ****** 
changed: [127.0.0.1]

TASK [Fetching phase from the chaos result CR] *********************************
2019-11-22T11:25:54.609941 (delta: 0.542445)         elapsed: 60.175438 ******* 
changed: [127.0.0.1]

TASK [debug] *******************************************************************
2019-11-22T11:25:55.141596 (delta: 0.531622)         elapsed: 60.707093 ******* 
ok: [127.0.0.1] => {
    "msg": "verdict : pass"
}

TASK [Create jinja template from chaosengine yaml] *****************************
2019-11-22T11:25:55.199871 (delta: 0.058178)         elapsed: 60.765368 ******* 
changed: [127.0.0.1]

TASK [patch experimentStatuses in jinja template for chaosengine] **************
2019-11-22T11:25:55.748500 (delta: 0.54842)         elapsed: 61.313997 ******** 
changed: [127.0.0.1] => (item=status:)
changed: [127.0.0.1] => (item=  experimentStatuses:)

TASK [patch status in jinja template for chaosengine] **************************
2019-11-22T11:25:56.158555 (delta: 0.409992)         elapsed: 61.724052 ******* 
changed: [127.0.0.1] => (item=  - name: {{experiment_name}})
changed: [127.0.0.1] => (item=    instance: {{job_name}})
changed: [127.0.0.1] => (item=    status: {{phase}})
changed: [127.0.0.1] => (item=    verdict: {{verdict}})

TASK [Patch the chaosengine status] ********************************************
2019-11-22T11:25:56.712881 (delta: 0.55427)         elapsed: 62.278378 ******** 
changed: [127.0.0.1]

TASK [Apply the chaos engine CR] ***********************************************
2019-11-22T11:25:57.148325 (delta: 0.435396)         elapsed: 62.713822 ******* 
changed: [127.0.0.1]

TASK [Display chaos-engine yaml] ***********************************************
2019-11-22T11:25:57.812181 (delta: 0.663811)         elapsed: 63.377678 ******* 
changed: [127.0.0.1]

TASK [debug] *******************************************************************
2019-11-22T11:25:58.344996 (delta: 0.532767)         elapsed: 63.910493 ******* 
ok: [127.0.0.1] => {
    "msg": "engine : apiVersion: litmuschaos.io/v1alpha1\nkind: ChaosEngine\nmetadata:\n  annotations:\n    kubectl.kubernetes.io/last-applied-configuration: |\n      {\"apiVersion\":\"litmuschaos.io/v1alpha1\",\"kind\":\"ChaosEngine\",\"metadata\":{\"annotations\":{},\"creationTimestamp\":\"2019-11-22T11:24:46Z\",\"generation\":1,\"name\":\"engine-nginx\",\"namespace\":\"default\",\"resourceVersion\":\"583719\",\"selfLink\":\"/apis/litmuschaos.io/v1alpha1/namespaces/default/chaosengines/engine-nginx\",\"uid\":\"be28c85d-6194-4ef9-ab48-5265c1debe88\"},\"spec\":{\"appinfo\":{\"appkind\":\"deployment\",\"applabel\":\"app=nginx\",\"appns\":\"default\"},\"chaosServiceAccount\":\"litmus\",\"experiments\":[{\"name\":\"pod-delete\",\"spec\":{\"components\":null}}],\"monitoring\":true,\"schedule\":{\"concurrencyPolicy\":\"\",\"excludedDays\":\"\",\"excludedTimes\":\"\",\"interval\":\"\"}},\"status\":{\"experimentStatuses\":[{\"instance\":\"pod-delete-9e95f5\",\"name\":\"pod-delete\",\"status\":\"\\u003cnil\\u003e\",\"verdict\":\"pass\"}]}}\n  creationTimestamp: 2019-11-22T11:24:46Z\n  generation: 2\n  name: engine-nginx\n  namespace: default\n  resourceVersion: \"583909\"\n  selfLink: /apis/litmuschaos.io/v1alpha1/namespaces/default/chaosengines/engine-nginx\n  uid: be28c85d-6194-4ef9-ab48-5265c1debe88\nspec:\n  appinfo:\n    appkind: deployment\n    applabel: app=nginx\n    appns: default\n  chaosServiceAccount: litmus\n  experiments:\n  - name: pod-delete\n    spec:\n      components: null\n  monitoring: true\n  schedule:\n    concurrencyPolicy: \"\"\n    excludedDays: \"\"\n    excludedTimes: \"\"\n    interval: \"\"\nstatus:\n  experimentStatuses:\n  - instance: pod-delete-9e95f5\n    name: pod-delete\n    status: <nil>\n    verdict: pass"
}

TASK [Get the jobCleanUpPolicy] ************************************************
2019-11-22T11:25:58.398887 (delta: 0.053829)         elapsed: 63.964384 ******* 
changed: [127.0.0.1]

TASK [Delete job if the jobCleanUpPolicy status is delete] *********************
2019-11-22T11:25:58.952741 (delta: 0.553743)         elapsed: 64.518238 ******* 
skipping: [127.0.0.1]

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=55   changed=30   unreachable=0    failed=0   

2019-11-22T11:25:58.974718 (delta: 0.021931)         elapsed: 64.540215 ******* 
===============================================================================

File: executor/executor.yml [hunk 2]
```diff
@@ -163,8 +170,8 @@
    executable: /bin/bash
   register: c_job_status
   until: c_job_status.stdout == "Complete"
-  delay: 1
-  retries: 300
+  delay: 10
+  retries: "{{ (chaos_monitoring_time.stdout)|int + 500}}"
```
Comment (ksatchit) [inline executor/executor.yml]: Just thought about this @amitbhatt818 . If our delay is >1 (of the tune of, say, 5-10ms), then we wouldn't need to add the additional default fixed seconds (500). As the {{ chaos_monitoring_time.stdout }} * {{ delay }} itself would be >> expected time for job completion. To understand better, let us take this example: 

TOTAL_CHAOS_DURATION: 600s

- With earlier case executor would terminate if job is still in progress
- With the new approach, max wait would be (600 + 500)* 10 -- 11000s (worst case/stuck jobs will need 3+ hours to recover/end.) 

I suppose an optimal estimation would be:

- If we decide to provide much larger window to account for post chaos stuck-ups (say every cleanup post chaos has a high wait interval), then ( {{ (chaos_monitoring_time.stdout) | int + 300s }} * 2s

**Pull request merged**